File: grix-egg/scripts/start_gateway.py
#!/usr/bin/env python3
"""Start and verify a Hermes gateway for one profile."""

from __future__ import annotations

import argparse
import json
import os
from pathlib import Path
import shutil
import subprocess
import sys
from typing import Any


NEGATIVE_STATUS_HINTS = [
    "not running",
    "not installed",
    "installed but not running",
    "inactive",
    "stopped",
]

POSITIVE_STATUS_HINTS = [
    "running",
    "healthy",
    "installed and running",
]


def clean_text(value: Any) -> str:
    return str(value or "").strip()


def resolve_hermes_home(explicit: str) -> Path:
    raw = clean_text(explicit) or clean_text(os.environ.get("HERMES_HOME")) or "~/.hermes"
    return Path(os.path.expanduser(raw)).resolve()


def resolve_profile_dir(hermes_home: Path, profile_name: str) -> Path:
    normalized = clean_text(profile_name)
    if normalized in {"", "default"}:
        return hermes_home
    return (hermes_home / "profiles" / normalized).resolve()


def ensure_hermes_binary(hermes_cmd: str) -> None:
    if os.sep in hermes_cmd:
        candidate = Path(os.path.expanduser(hermes_cmd)).resolve()
        if not candidate.exists():
            raise RuntimeError(f"Hermes CLI not found: {candidate}")
        return
    if shutil.which(hermes_cmd):
        return
    raise RuntimeError(
        f"Hermes CLI '{hermes_cmd}' is not available in PATH. "
        "Install Hermes first or pass --hermes with an absolute path."
    )


def profile_prefix(hermes_cmd: str, profile_name: str) -> list[str]:
    normalized = clean_text(profile_name)
    if normalized in {"", "default"}:
        return [hermes_cmd]
    return [hermes_cmd, "--profile", normalized]


def run_command(cmd: list[str], env: dict[str, str], check: bool = True) -> subprocess.CompletedProcess[str]:
    result = subprocess.run(cmd, text=True, capture_output=True, env=env)
    if check and result.returncode != 0:
        stderr = (result.stderr or "").strip()
        stdout = (result.stdout or "").strip()
        raise RuntimeError(stderr or stdout or f"command failed: {' '.join(cmd)}")
    return result


def summarize_output(result: subprocess.CompletedProcess[str]) -> str:
    parts = [clean_text(result.stdout), clean_text(result.stderr)]
    return "\n".join(part for part in parts if part)


def status_is_running(result: subprocess.CompletedProcess[str]) -> bool:
    if result.returncode != 0:
        return False
    combined = summarize_output(result).lower()
    if not combined:
        return False
    if any(hint in combined for hint in NEGATIVE_STATUS_HINTS):
        return False
    return any(hint in combined for hint in POSITIVE_STATUS_HINTS)


def main() -> int:
    parser = argparse.ArgumentParser(description="Start a Hermes gateway and verify it is running.")
    parser.add_argument("--profile-name", default="")
    parser.add_argument("--hermes-home", default="")
    parser.add_argument("--hermes", default="hermes")
    parser.add_argument("--start-subcommand", choices=["start", "run"], default="start")
    parser.add_argument("--status-subcommand", default="status")
    parser.add_argument("--json", action="store_true")
    args = parser.parse_args()

    try:
        hermes_home = resolve_hermes_home(args.hermes_home)
        profile_name = clean_text(args.profile_name)
        profile_dir = resolve_profile_dir(hermes_home, profile_name)
        if not profile_dir.exists():
            raise RuntimeError(f"Hermes profile does not exist: {profile_dir}")

        ensure_hermes_binary(args.hermes)
        env = dict(os.environ)
        env["HERMES_HOME"] = str(hermes_home)

        command_prefix = profile_prefix(args.hermes, profile_name)
        status_cmd = command_prefix + ["gateway", args.status_subcommand]
        status_before = run_command(status_cmd, env=env, check=False)
        already_running = status_is_running(status_before)

        start_result: subprocess.CompletedProcess[str] | None = None
        if not already_running:
            start_cmd = command_prefix + ["gateway", args.start_subcommand]
            start_result = run_command(start_cmd, env=env, check=False)
            if start_result.returncode != 0:
                raise RuntimeError(
                    "Failed to start Hermes gateway.\n"
                    f"command: {' '.join(start_cmd)}\n"
                    f"output:\n{summarize_output(start_result)}"
                )

        status_after = run_command(status_cmd, env=env, check=False)
        if not status_is_running(status_after):
            raise RuntimeError(
                "Hermes gateway did not report a running state after startup.\n"
                f"command: {' '.join(status_cmd)}\n"
                f"output:\n{summarize_output(status_after)}"
            )

        payload = {
            "ok": True,
            "profile_name": profile_name or "default",
            "hermes_home": str(hermes_home),
            "profile_dir": str(profile_dir),
            "already_running": already_running,
            "start_subcommand": args.start_subcommand,
            "status_before": {
                "code": status_before.returncode,
                "stdout": clean_text(status_before.stdout),
                "stderr": clean_text(status_before.stderr),
            },
            "status_after": {
                "code": status_after.returncode,
                "stdout": clean_text(status_after.stdout),
                "stderr": clean_text(status_after.stderr),
            },
            "start_result": None if start_result is None else {
                "code": start_result.returncode,
                "stdout": clean_text(start_result.stdout),
                "stderr": clean_text(start_result.stderr),
            },
        }
        if args.json:
            print(json.dumps(payload, ensure_ascii=False, indent=2))
        else:
            print(json.dumps(payload, ensure_ascii=False))
        return 0
    except Exception as exc:  # noqa: BLE001
        payload = {"ok": False, "error": str(exc)}
        if args.json:
            print(json.dumps(payload, ensure_ascii=False, indent=2), file=sys.stderr)
        else:
            print(str(exc), file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: grix-egg/scripts/start_gateway.ts
#!/usr/bin/env node
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawn, spawnSync } from "node:child_process";

const NEGATIVE_STATUS_HINTS = [
  "not running",
  "not installed",
  "installed but not running",
  "inactive",
  "stopped",
];

const POSITIVE_STATUS_HINTS = [
  "running",
  "healthy",
  "installed and running",
];

type StartSubcommand = "start" | "run";

interface Flags {
  profileName: string;
  hermesHome: string;
  hermes: string;
  startSubcommand: StartSubcommand;
  statusSubcommand: string;
  json: boolean;
}

interface CommandOutput {
  code: number;
  stdout: string;
  stderr: string;
}

interface StartAttempt {
  mode: "already_running" | "service_start" | "service_install_start" | "manual_run_detached";
  command: string[];
  result: CommandOutput;
}

function cleanText(value: unknown): string {
  return String(value ?? "").trim();
}

function expandHome(value: string): string {
  if (!value) return value;
  if (value === "~") return os.homedir();
  if (value.startsWith("~/")) return path.join(os.homedir(), value.slice(2));
  return value;
}

function resolveHermesHome(explicit: string): string {
  const raw = cleanText(explicit) || cleanText(process.env.HERMES_HOME) || "~/.hermes";
  return path.resolve(expandHome(raw));
}

function resolveProfileDir(hermesHome: string, profileName: string): string {
  const normalized = cleanText(profileName);
  if (!normalized || normalized === "default") return hermesHome;
  return path.resolve(path.join(hermesHome, "profiles", normalized));
}

function ensureHermesBinary(hermesCmd: string): void {
  if (hermesCmd.includes(path.sep)) {
    const candidate = path.resolve(expandHome(hermesCmd));
    if (!fs.existsSync(candidate)) {
      throw new Error(`Hermes CLI not found: ${candidate}`);
    }
    return;
  }
  const result = spawnSync("which", [hermesCmd], { encoding: "utf8" });
  if ((result.status ?? -1) !== 0) {
    throw new Error(
      `Hermes CLI '${hermesCmd}' is not available in PATH. ` +
        "Install Hermes first or pass --hermes with an absolute path.",
    );
  }
}

function profilePrefix(hermesCmd: string, profileName: string): string[] {
  const normalized = cleanText(profileName);
  if (!normalized || normalized === "default") return [hermesCmd];
  return [hermesCmd, "--profile", normalized];
}

function runCommand(cmd: string[], env: NodeJS.ProcessEnv, check = true): CommandOutput {
  const [bin, ...rest] = cmd;
  if (!bin) throw new Error("runCommand received empty cmd");
  const result = spawnSync(bin, rest, { encoding: "utf8", env });
  const output: CommandOutput = {
    code: result.status ?? -1,
    stdout: (result.stdout || "").trim(),
    stderr: (result.stderr || "").trim(),
  };
  if (check && output.code !== 0) {
    throw new Error(output.stderr || output.stdout || `command failed: ${cmd.join(" ")}`);
  }
  return output;
}

function runDetachedCommand(cmd: string[], env: NodeJS.ProcessEnv): CommandOutput {
  const [bin, ...rest] = cmd;
  if (!bin) throw new Error("runDetachedCommand received empty cmd");
  const child = spawn(bin, rest, {
    detached: true,
    env,
    stdio: "ignore",
  });
  child.unref();
  return {
    code: 0,
    stdout: `detached pid=${child.pid ?? ""}`.trim(),
    stderr: "",
  };
}

function summarizeOutput(result: CommandOutput): string {
  return [cleanText(result.stdout), cleanText(result.stderr)]
    .filter(Boolean)
    .join("\n");
}

function statusIsRunning(result: CommandOutput): boolean {
  if (result.code !== 0) return false;
  const combined = summarizeOutput(result).toLowerCase();
  if (!combined) return false;
  if (NEGATIVE_STATUS_HINTS.some((hint) => combined.includes(hint))) return false;
  return POSITIVE_STATUS_HINTS.some((hint) => combined.includes(hint));
}

function serviceLooksUnavailable(result: CommandOutput): boolean {
  const combined = summarizeOutput(result).toLowerCase();
  return (
    result.code !== 0 &&
    (
      combined.includes("not installed") ||
      combined.includes("service is not installed") ||
      combined.includes("no such file") ||
      combined.includes("could not find") ||
      combined.includes("not loaded") ||
      combined.includes("unloaded")
    )
  );
}

function waitForRunning(statusCmd: string[], env: NodeJS.ProcessEnv): CommandOutput {
  let latest = runCommand(statusCmd, env, false);
  const deadline = Date.now() + 10000;
  while (!statusIsRunning(latest) && Date.now() < deadline) {
    Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, 500);
    latest = runCommand(statusCmd, env, false);
  }
  return latest;
}

function parseArgs(argv: string[]): Flags {
  const flags: Flags = {
    profileName: "",
    hermesHome: "",
    hermes: "hermes",
    startSubcommand: "start",
    statusSubcommand: "status",
    json: false,
  };
  for (let i = 0; i < argv.length; i += 1) {
    const token = argv[i]!;
    const next = argv[i + 1];
    if (token === "--profile-name" && next !== undefined) { flags.profileName = next; i += 1; continue; }
    if (token === "--hermes-home" && next !== undefined) { flags.hermesHome = next; i += 1; continue; }
    if (token === "--hermes" && next !== undefined) { flags.hermes = next; i += 1; continue; }
    if (token === "--start-subcommand" && next !== undefined) {
      if (next !== "start" && next !== "run") throw new Error(`Invalid --start-subcommand: ${next}`);
      flags.startSubcommand = next as StartSubcommand;
      i += 1;
      continue;
    }
    if (token === "--status-subcommand" && next !== undefined) { flags.statusSubcommand = next; i += 1; continue; }
    if (token === "--json") { flags.json = true; continue; }
  }
  return flags;
}

function main(): number {
  let flags: Flags;
  try {
    flags = parseArgs(process.argv.slice(2));
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    process.stderr.write(`${message}\n`);
    return 1;
  }

  try {
    const hermesHome = resolveHermesHome(flags.hermesHome);
    const profileName = cleanText(flags.profileName);
    const profileDir = resolveProfileDir(hermesHome, profileName);
    if (!fs.existsSync(profileDir)) {
      throw new Error(`Hermes profile does not exist: ${profileDir}`);
    }

    ensureHermesBinary(flags.hermes);
    const env = { ...process.env, HERMES_HOME: hermesHome };

    const commandPrefix = profilePrefix(flags.hermes, profileName);
    const statusCmd = [...commandPrefix, "gateway", flags.statusSubcommand];
    const statusBefore = runCommand(statusCmd, env, false);
    const alreadyRunning = statusIsRunning(statusBefore);

    let startAttempt: StartAttempt | null = null;
    if (!alreadyRunning) {
      const startCmd = [...commandPrefix, "gateway", flags.startSubcommand];
      const startResult = runCommand(startCmd, env, false);
      startAttempt = { mode: "service_start", command: startCmd, result: startResult };

      if (startResult.code !== 0 && flags.startSubcommand === "start" && serviceLooksUnavailable(startResult)) {
        const installCmd = [...commandPrefix, "gateway", "install"];
        const installResult = runCommand(installCmd, env, false);
        if (installResult.code === 0) {
          const retryResult = runCommand(startCmd, env, false);
          startAttempt = {
            mode: "service_install_start",
            command: startCmd,
            result: retryResult,
          };
        }
      }

      if (startAttempt.result.code !== 0) {
        const runCmd = [...commandPrefix, "gateway", "run"];
        startAttempt = {
          mode: "manual_run_detached",
          command: runCmd,
          result: runDetachedCommand(runCmd, env),
        };
      }
    }

    let statusAfter = waitForRunning(statusCmd, env);
    if (!statusIsRunning(statusAfter) && startAttempt && startAttempt.mode !== "manual_run_detached") {
      const runCmd = [...commandPrefix, "gateway", "run"];
      startAttempt = {
        mode: "manual_run_detached",
        command: runCmd,
        result: runDetachedCommand(runCmd, env),
      };
      statusAfter = waitForRunning(statusCmd, env);
    }
    if (!statusIsRunning(statusAfter)) {
      throw new Error(
        "Hermes gateway did not report a running state after startup.\n" +
          `command: ${statusCmd.join(" ")}\n` +
          `start_mode: ${startAttempt?.mode || "already_running"}\n` +
          `start_output:\n${startAttempt ? summarizeOutput(startAttempt.result) : ""}\n` +
          `status_output:\n${summarizeOutput(statusAfter)}`,
      );
    }

    const payload = {
      ok: true as const,
      profile_name: profileName || "default",
      hermes_home: hermesHome,
      profile_dir: profileDir,
      already_running: alreadyRunning,
      start_subcommand: flags.startSubcommand,
      start_mode: startAttempt?.mode || "already_running",
      status_before: statusBefore,
      status_after: statusAfter,
      start_result: startAttempt?.result || null,
      start_command: startAttempt?.command || null,
    };
    if (flags.json) {
      process.stdout.write(`${JSON.stringify(payload, null, 2)}\n`);
    } else {
      process.stdout.write(`${JSON.stringify(payload)}\n`);
    }
    return 0;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    const payload = { ok: false, error: message };
    if (flags.json) {
      process.stderr.write(`${JSON.stringify(payload, null, 2)}\n`);
    } else {
      process.stderr.write(`${message}\n`);
    }
    return 1;
  }
}

process.exit(main());
